gwpy/timeseries/io/core.py

```diff
@@ -56,11 +63,42 @@ def _join(data):
                 tsd = data.pop(0)
                 out.append(tsd, gap=gap, pad=pad)
                 del tsd
+            for key in out:
+                out[key] = _pad_series(
+                    out[key],
+                    pad,
+                    start,
+                    end,
+                )
             return out
     else:
         from .. import TimeSeriesBaseList
 
         def _join(arrays):
             list_ = TimeSeriesBaseList(*arrays)
-            return list_.join(pad=pad, gap=gap)
+            return _pad_series(
+                list_.join(pad=pad, gap=gap),
+                pad,
+                start,
+                end,
+            )
     return _join
+
+
+def _pad_series(ts, pad, start=None, end=None):
+    """Pad a timeseries to match the specified [start, end) limits
+
+    To cover a gap in data returned from a data source
+    """
+    span = ts.span
+    if start is None:
```

Comment:
@duncanmmacleod, is it worth trying to define `pada` and `padb` on one line each?

```python
    pada = (0 if start is None else
            max(int((span[0] - start) * ts.sample_rate.value), 0))
    padb = (0 if end is None else
            max(int((end - span[1]) * ts.sample_rate.value), 0))
```

Personally, I just think the latter looks a little cleaner.

Reply:
I have simplified the logic in a different way, please re-review.

Reply:
Looks good, we're really cookin' with gas now.